File: mypay.sources/mypay4-be/src/main/java/it/regioneveneto/mygov/payment/mypay4/service/AvvisoService.java
/**
 *     MyPay - Payment portal of Regione Veneto.
 *     Copyright (C) 2022  Regione Veneto
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package it.regioneveneto.mygov.payment.mypay4.service;

import it.regioneveneto.mygov.payment.mypay4.dao.AvvisoDao;
import it.regioneveneto.mygov.payment.mypay4.dto.AnagraficaPagatore;
import it.regioneveneto.mygov.payment.mypay4.model.Avviso;
import it.regioneveneto.mygov.payment.mypay4.model.Ente;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Service
@Slf4j
@Transactional(propagation = Propagation.SUPPORTS)
public class AvvisoService {

  @Autowired
  private AvvisoDao avvisoDao;

  @Transactional(propagation = Propagation.REQUIRED)
  public Avviso insert(Ente ente, String codIuv, AnagraficaPagatore anagraficaPagatore) {
    Avviso avviso = new Avviso();

    avviso.setCodIuv(codIuv);
    avviso.setMygovEnteId(ente);

    avviso.setCodRpSoggPagIdUnivPagTipoIdUnivoco(Character.toString(anagraficaPagatore.getTipoIdentificativoUnivoco()));
    avviso.setCodRpSoggPagIdUnivPagCodiceIdUnivoco(anagraficaPagatore.getCodiceIdentificativoUnivoco());
    avviso.setDeRpSoggPagAnagraficaPagatore(anagraficaPagatore.getAnagrafica());
    avviso.setDeRpSoggPagEmailPagatore(anagraficaPagatore.getEmail());

    avviso.setDeRpSoggPagIndirizzoPagatore(anagraficaPagatore.getIndirizzo());
    avviso.setDeRpSoggPagCivicoPagatore(anagraficaPagatore.getCivico());
    avviso.setCodRpSoggPagCapPagatore(anagraficaPagatore.getCap());

    avviso.setDeRpSoggPagLocalitaPagatore(
        StringUtils.isBlank(anagraficaPagatore.getLocalita()) ? null : anagraficaPagatore.getLocalita());
    avviso.setDeRpSoggPagProvinciaPagatore(
        StringUtils.isBlank(anagraficaPagatore.getProvincia()) ? null : anagraficaPagatore.getProvincia());
    avviso.setCodRpSoggPagNazionePagatore(
        StringUtils.isBlank(anagraficaPagatore.getNazione()) ? null : anagraficaPagatore.getNazione());

    long mygovAvvisoId = avvisoDao.insert(avviso);
    avviso.setMygovAvvisoId(mygovAvvisoId);
    return avviso;
  }

  public Optional<Avviso> getByIuvEnte(final String codIuv, final String codIpaEnte) {
    return avvisoDao.getByIuvEnte(codIuv, codIpaEnte);
  }
}

File: mypay.sources/mypay4-fe/projects/mypay4-fe-cittadino/src/app/services/avviso.service.ts
/*
 *     MyPay - Payment portal of Regione Veneto.
 *     Copyright (C) 2022  Regione Veneto
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import {
    ApiInvokerService, BaseUrlService, mapper
} from 'projects/mypay4-fe-common/src/public-api';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Debito } from '../model/debito';
import { Pagato } from '../model/pagato';

@Injectable({
  providedIn: 'root'
})
export class AvvisoService {

  constructor(
    private baseUrlService: BaseUrlService,
    private apiInvokerService: ApiInvokerService
  ) { }

  searchAvvisi(numeroAvviso: string, owner: boolean, codIdUnivoco: string, anagrafica: string): Observable<any> {
    return this.searchAvvisiImpl(false, numeroAvviso, owner, codIdUnivoco, anagrafica, null);
  }

  searchAvvisiAnonymous(numeroAvviso: string, codIdUnivoco: string, anagrafica: string, recaptchaResponse: any): Observable<any> {
    return this.searchAvvisiImpl(true, numeroAvviso, null, codIdUnivoco, anagrafica, recaptchaResponse);
  }

  private searchAvvisiImpl(anonymous:boolean, numeroAvviso: string, owner: boolean, codIdUnivoco: string, anagrafica: string, recaptchaResponse: any): Observable<any> {
    let params = new HttpParams();
    params = params.append('numeroAvviso', numeroAvviso);
    if(!owner) {
      if(codIdUnivoco)
        params = params.append('codIdUnivoco', codIdUnivoco);
      if(anagrafica)
        params = params.append('anagrafica', anagrafica);
    }
    if(anonymous){
      params = params.append('recaptcha', recaptchaResponse);
    } else {
      params = params.append('owner', String(owner));
    }

    return this.apiInvokerService.get<any>(this.baseUrlService.getCittadinoUrl(anonymous) + 'avvisi/search', {params:params})
      .pipe(map(response => {
        response.debiti = response.debiti?.map(item => mapper(item, Debito.MAPPER_S2C_DEF));
        response.pagati = response.pagati?.map(item => mapper(item, Pagato.MAPPER_S2C_DEF));
        return response;
      }));
  }

  downloadAvviso(debito: Debito): any {
    return this.downloadAvvisoImpl(false, debito);
  }

  downloadAvvisoAnonymous(debito: Debito, recaptchaResponse: any): any {
    return this.downloadAvvisoImpl(true, debito, recaptchaResponse);
  }

  private downloadAvvisoImpl(anonymous:boolean, debito: Debito, recaptchaResponse?: any): any {
    let params = new HttpParams();
    if(anonymous)
      params = params.append('recaptcha', recaptchaResponse);
    if(debito.securityTokenAvviso)
      params = params.append('securityToken', debito.securityTokenAvviso);
    return this.apiInvokerService.get<any>(this.baseUrlService.getCittadinoUrl(anonymous) + 'avvisi/download/' + debito.id, {params:params, observe: 'response', responseType: 'blob'});
  }

}
